sacred/config/custom_containers.py

```diff
@@ -300,6 +300,10 @@ def type_changed(old_value, new_value):
 def is_different(old_value, new_value):
     """Numpy aware comparison between two values."""
     if opt.has_numpy:
-        return not opt.np.array_equal(old_value, new_value)
-    else:
-        return old_value != new_value
+        # np.array_equal raises an exception when the arguments are not array in numpy 2.0.
+        # This issue is only present in 2.0, not in <2.0 or >=2.1
+        if isinstance(old_value, opt.np.ndarray) and isinstance(new_value, opt.np.ndarray):
+            return not opt.np.array_equal(old_value, new_value)
+        elif isinstance(old_value, opt.np.ndarray) or isinstance(new_value, opt.np.ndarray):
+            return False
```

Comment:
This check is not identical to the previous check. The issue is https://github.com/numpy/numpy/issues/27271

Reply:
This function should do it
```python
def np_array_equal(a1, a2, equal_nan: bool = False):
    # Adapted version of np.array_equal that avoids
    # https://github.com/numpy/numpy/issues/27271
    assert opt.has_numpy, "numpy is not available"
    try:
        a1, a2 = opt.np.asarray(a1), opt.np.asarray(a2)
    except Exception:
        return False
    if a1.shape != a2.shape:
        return False
    if not equal_nan:
        result = a1 == a2
        if isinstance(result, bool):
            return result
        return builtins.bool(result.all())
    cannot_have_nan = opt.np._dtype_cannot_hold_nan(
        a1.dtype
    ) and opt.np._dtype_cannot_hold_nan(a2.dtype)
    if cannot_have_nan:
        if a1 is a2:
            return True
        result = a1 == a2
        if isinstance(result, bool):
            return result
        return builtins.bool(result.all())

    if a1 is a2:
        # nan will compare equal so an array will compare equal to itself.
        return True
    # Handling NaN values if equal_nan is True
    a1nan, a2nan = opt.np.isnan(a1), opt.np.isnan(a2)
    # NaN's occur at different locations
    if not (a1nan == a2nan).all():
        return False
    # Shapes of a1, a2 and masks are guaranteed to be consistent by this point
    # Here we don't need to check that the result might be a boolean, because it doesn't
    # happen for numerical values.
    return builtins.bool((a1[~a1nan] == a2[~a1nan]).all())
```

Reply:
I copied this function from `np.array_equal` and added checks where necessary to avoid https://github.com/numpy/numpy/issues/27271

Reply:
Well, GitHub didn't display your other messages... 

Since `equal_nan=False`, we can get away with just a few lines:

```py
        try:
            old_value = opt.np.asarray(old_value)
            new_value = opt.np.asarray(new_value)
        except:
            return False
        else:
            result = opt.np.asarray(old_value) == opt.np.asarray(new_value)
            if isinstance(result, bool):
                return result
            else:
                return result.all()
``` 

But I found another issue with the original code. It does change behavior in some cases when numpy is present vs when not:

```py
>>> [1, [1,2]] == [1, [1,2]]
True
>>> np.array_equal([1, [1,2]], [1, [1,2]])
False
```

Which raises the question if we should really keep the old behavior...

Reply:
This looks very undesired and potentially undesired tbh. I haven't looked deeply into what code paths would be affected by fixing this.

Reply:
The old behavior should be recovered in numpy 2.0.1 and 2.1.1 https://github.com/numpy/numpy/issues/27271
I don't know when that will be published but that should pretty much eliminate the need to write code that restores the old behavior. I guess the important question is whether one should fix the old behavior.

Reply:
I'll open an issue for this and reproduce the old behavior in this PR. I'm not sure yet what impact it would have to fix this. I guess that most sacred users have numpy installed so that fixing this may actually break or change something for them.

Reply:
sounds reasonable. However, iirc I had some issues with the identity check once, and the solution was counterintuitive. But, that may have been `seml` and not `sacred`'s fault. Also, it has been years...